### PR TITLE
Closes #7 : Octal literals

### LINKED ISSUE #7: Parse octal number literals
quick-lint-js understands hexadecimal, binary, and decimal number literals, but doesn't understand octal number literals.

Teach quick-lint-js to parse octal number literals in lex.cpp and in test-lex.cpp.

### PR DESCRIPTION
**Pull request by vegerot**
Left some questions in comments as I was going through it.  Some of the question are obsolete

### CODE REVIEW

**Comment by vegerot** [comment]
Damn.  Spent 20 minutes rebasing before giving up and merging instead.  Did not realize this PR would be rebased.

File: .gitignore [hunk 1]
```diff
@@ -1 +1,22 @@
+.clangd/
```
Comment (vegerot) [inline .gitignore:1]: Sue me

File: src/lex.cpp [hunk 2]
```diff
@@ -144,16 +147,25 @@ void lexer::parse_current_token() {
...
           case 'b':
           case 'B':
             this->input_ += 2;
             this->parse_binary_number();
             break;
+          case 'o':
+            this->input_ += 2;
+            this->parse_octal_number(true);
+            break;
+            // probably wrong
```
Comment (vegerot) [inline src/lex.cpp:159]: probably not

File: src/lex.cpp [hunk 3]
```diff
@@ -634,34 +657,55 @@ void lexer::parse_binary_number() {
...
+        this->input_ = input;
+        this->parse_number();
+        return;
+      }
     QLJS_CASE_IDENTIFIER_START:
-      input += 1;
-      break;
+    case u8'.':
+      ++input;
+      goto done_parsing_garbage;
```
Comment (vegerot) [inline src/lex.cpp:684]: Maybe this isn't the right thing.  But the issue was if I had 

```suggestion
      break;
```

a decimal like `012.3` would see the `.`, break, increment, then the `3` would be seen by [678](https://github.com/strager/quick-lint-js/pull/40/commits/d44727a46e5cca624bcc21f421ca05c5bb331571#diff-3f07da05c7d34d6ae43baceb4a820032R684) and the execution disappears forever.

File: src/lex.cpp [hunk 4]
```diff
@@ -634,34 +657,55 @@ void lexer::parse_binary_number() {
...
       default:
         goto done_parsing_garbage;
     }
   }
+  /* note(👮🏾‍♀️)this is bad, because it won't parse the trailing
+   * garbage separate tokens correctly. for example, `0o57.2` should be two
+   * number tokens, while `0o5a` should be an "unknown" token and an "any" token
+   */
```
Comment (vegerot) [inline src/lex.cpp:692]: This was back when I was narrow-mindedly focused on V8, instead of realizing the freedom we could take

```suggestion
```

File: src/lex.cpp [hunk 5]
```diff
@@ -634,34 +657,55 @@ void lexer::parse_binary_number() {
...
-  return input;
+  this->input_ = input;
 }
 
 void lexer::parse_number() {
   QLJS_ASSERT(this->is_digit(this->input_[0]) || this->input_[0] == '.');
   const char8* number_begin = this->input_;
   const char8* input = number_begin;
 
+  // @@@ (👮🏾‍♀️)can I use this other places?
```
Comment (vegerot) [inline src/lex.cpp:708]: Already done in https://github.com/strager/quick-lint-js/commit/4ecd4c69ec86c0f97b5a1af389796f2fc4693c5e

But unfortunately won't use it in all of my implementations

```suggestion
```

File: test/test-lex.cpp [hunk 6]
```diff
@@ -118,12 +118,109 @@ TEST(test_lex, lex_binary_numbers) {
...
+  check_single_token(u8"051", token_type::number);
+  check_single_token(u8"0o51", token_type::number);
+}
+
+TEST(test_lex, lex_octal_numbers_lax) {
+  check_single_token(u8"058", token_type::number);
+  check_single_token(u8"058.9", token_type::number);
+}
+
+// TODO(👮🏾‍♀️)doesn't work
```
Comment (vegerot) [inline test/test-lex.cpp:153]: Does work, though I wish better

```suggestion
```

File: test/test-lex.cpp [hunk 7]
```diff
@@ -118,12 +118,109 @@ TEST(test_lex, lex_binary_numbers) {
...
+    lexer l(&input, &v);
+    EXPECT_EQ(l.peek().type, token_type::number);
+    l.skip();
+    EXPECT_EQ(l.peek().type, token_type::number);
+    l.skip();
+    EXPECT_EQ(l.peek().type, token_type::end_of_file);
+
+    ASSERT_EQ(v.errors.size(), 1);
+    // q(🎅🏾) have another error kind for
+    // `unexpected_character_in_octal_number?
```
Comment (vegerot) [inline test/test-lex.cpp:185]:
```suggestion
```

File: test/test-lex.cpp [hunk 8]
```diff
@@ -118,12 +118,109 @@ TEST(test_lex, lex_binary_numbers) {
...
+  {
+    error_collector v;
+    padded_string input(u8"0o58");
+    lexer l(&input, &v);
+    EXPECT_EQ(l.peek().type, token_type::number);
+    l.skip();
+    EXPECT_EQ(l.peek().type, token_type::end_of_file);
+
+    // q(🎅🏾) have another error kind for
+    // `unexpected_character_in_octal_number?
```
Comment (vegerot) [inline test/test-lex.cpp:168]:
```suggestion
```

File: test/test-lex.cpp [hunk 9]
```diff
@@ -118,12 +118,109 @@ TEST(test_lex, lex_binary_numbers) {
...
+    error_collector v;
+    padded_string input(u8"0o58.2");
+    lexer l(&input, &v);
+    EXPECT_EQ(l.peek().type, token_type::number);
+    l.skip();
+    EXPECT_EQ(l.peek().type, token_type::number);
+    l.skip();
+    EXPECT_EQ(l.peek().type, token_type::end_of_file);
+
+    ASSERT_EQ(v.errors.size(), 1);
```
Comment (vegerot) [inline test/test-lex.cpp:183]: Unnecessary 

```suggestion
```

File: test/test-lex.cpp [hunk 10]
```diff
@@ -118,12 +118,109 @@ TEST(test_lex, lex_binary_numbers) {
...
+// TODO (👮🏾‍♀️) (when strict mode implemented) tests to fail in
+// strict mode
+
 TEST(test_lex, lex_hex_numbers) {
   check_single_token(u8"0x0", token_type::number);
   check_single_token(u8"0x123456789abcdef", token_type::number);
   check_single_token(u8"0X123456789ABCDEF", token_type::number);
 }
 
+TEST(test_lex, DISABLED_fail_lex_hex_numbers) {
```
Comment (vegerot) [inline test/test-lex.cpp:218]: @sTRAGER request for comment

File: test/test-lex.cpp [hunk 11]
```diff
@@ -208,6 +305,20 @@ TEST(test_lex, lex_number_with_trailing_garbage) {
                               error_unexpected_characters_in_number, characters,
                               offsets_matcher(&input, 4, 7))));
   }
+
+  // TODO(👮🏾‍♀️)doesn't work
```
Comment (vegerot) [inline test/test-lex.cpp:309]:
```suggestion
```

File: test/test-lex.cpp [hunk 12]
```diff
@@ -266,36 +377,50 @@ TEST(test_lex, lex_invalid_big_int_number) {
   }
 
   // Complain about both the decimal point and the leading 0 digit.
+  // q(👮🏽‍♀️)S
+  // o do we want this to be?  Should it complain it's a bad octal number
+  // (because of `.`) and bad big_int because of `.2n`?
+  // The problem with that is I lex the `.` in `lexer::parse_octal_number`
```
Comment (vegerot) [inline test/test-lex.cpp:383]: @sTRAGER request for comment

File: test/test-lex.cpp [hunk 13]
```diff
@@ -266,36 +377,50 @@ TEST(test_lex, lex_invalid_big_int_number) {
...
+    l.skip();
     EXPECT_EQ(l.peek().type, token_type::end_of_file);
 
     EXPECT_THAT(
         v.errors,
         UnorderedElementsAre(
-            VariantWith<error_big_int_literal_contains_decimal_point>(_),
-            VariantWith<error_big_int_literal_contains_leading_zero>(_)));
+            VariantWith<error_unexpected_characters_in_octal_number>(_)
+            /* VariantWith<error_big_int_literal_contains_decimal_point>(_), */
```
Comment (vegerot) [inline test/test-lex.cpp:398]: Wasn't sure what to do here

File: test/test-lex.cpp [hunk 14]
```diff
@@ -266,36 +377,50 @@ TEST(test_lex, lex_invalid_big_int_number) {
...
     EXPECT_EQ(l.peek().type, token_type::end_of_file);
 
     EXPECT_THAT(
         v.errors,
         UnorderedElementsAre(
-            VariantWith<error_big_int_literal_contains_decimal_point>(_),
-            VariantWith<error_big_int_literal_contains_exponent>(_),
-            VariantWith<error_big_int_literal_contains_leading_zero>(_)));
+            VariantWith<error_unexpected_characters_in_octal_number>(_),
+            /* VariantWith<error_big_int_literal_contains_decimal_point>(_), */
```
Comment (vegerot) [inline test/test-lex.cpp:422]: Not satisfied with this.  The test description says it should "Complain about everything", but I am getting rid of complaints

File: src/lex.cpp [hunk 15]
```diff
@@ -634,27 +657,43 @@ void lexer::parse_binary_number() {
...
+    ++input;
+  }
+
   const char8* garbage_begin = input;
   for (;;) {
     switch (*input) {
+    /* TODO (👮🏾‍♀️ mswitch switch to just 8 and 9)*/
+    // decimal point should also be included
     QLJS_CASE_DECIMAL_DIGIT:
+      if (strict) {
```
Comment (vegerot) [inline src/lex.cpp:673]: Don't check `strict` every time

File: test/test-lex.cpp [hunk 16]
```diff
@@ -118,12 +118,109 @@ TEST(test_lex, lex_binary_numbers) {
   check_single_token(u8"0B010101010101010", token_type::number);
 }
 
+// fails
```
Comment (vegerot) [inline test/test-lex.cpp:121]: I added this test because I found a potential for a bug while reading the binary lexer.  Didn't fix it, just wrote the test.  Maybe I should disable this because the test fails

File: src/lex.cpp (hunk 3 above)
Comment (vegerot) [reply src/lex.cpp:688]: May be resolved as of bf6e92bb6a2fa9087392459c181032234b210985

File: .gitignore (hunk 1 above)
Comment (strager) [reply .gitignore:1]: OBJECTION

Please undo your changes to this file.

File: src/include/quick-lint-js/error.h [hunk 17]
```diff
@@ -205,6 +205,11 @@
       error_unexpected_control_character, { source_code_span character; },     \
       .error(u8"unexpected control character", character))                     \
                                                                                \
+  QLJS_ERROR_TYPE(                                                             \
```
Comment (strager) [inline src/include/quick-lint-js/error.h:208]: Nit: Order alphabetically.

File: src/include/quick-lint-js/lex.h [hunk 18]
```diff
@@ -302,6 +302,7 @@ class lexer {
                                            error_reporter*);
 
   void parse_binary_number();
+  void parse_octal_number(bool strict, bool is_0o);
```
Comment (strager) [inline src/include/quick-lint-js/lex.h:305]: Instead of bool parameters, how about an enum? Maybe something like this:

```cpp
enum class octal_kind {
  non_strict,
  //strict_0,
  strict_0o,
};
```

File: src/lex.cpp [hunk 19]
```diff
@@ -142,16 +145,25 @@ void lexer::parse_current_token() {
...
         case 'b':
         case 'B':
           this->input_ += 2;
           this->parse_binary_number();
           break;
+        case 'o':
+          this->input_ += 2;
+          this->parse_octal_number(true, true);
+          break;
+          // probably wrong
```
Comment (strager) [inline src/lex.cpp:157]: Remove this comment.

File: src/lex.cpp [hunk 20]
```diff
@@ -722,16 +734,37 @@ const char8* lexer::end_of_previous_token() const noexcept {
 }
 
 void lexer::parse_hexadecimal_number() {
-  QLJS_ASSERT(this->is_hex_digit(this->input_[0]) || this->input_[0] == '.');
```
Comment (strager) [inline src/lex.cpp:725]: This change seems unrelated to the rest of your changes. I filed issue #43 to fix this problem. Undo?

File: src/lex.cpp [hunk 21]
```diff
@@ -722,16 +734,37 @@ const char8* lexer::end_of_previous_token() const noexcept {
...
   char8* input = this->input_;
 
   while (this->is_hex_digit(*input)) {
     input += 1;
   }
-
```
Comment (strager) [inline src/lex.cpp:731]: Nit: Undo changes.

File: src/lex.cpp [hunk 22]
```diff
@@ -722,16 +734,37 @@ const char8* lexer::end_of_previous_token() const noexcept {
...
   this->input_ = check_garbage_in_number_literal(input);
 }
 
+char8* lexer::check_garbage_in_number_literal(char8* input) {
```
Comment (strager) [inline src/lex.cpp:745]: Minor: Un-move this function so the Git-diff is easier to read.

File: src/lex.cpp [hunk 23]
```diff
@@ -743,27 +776,52 @@ void lexer::parse_binary_number() {
...
+      this->parse_number();
+      return;
+    }
+  }
+
+  if (is_0o) switch (*input)
+    case u8'n':
+      ++input;
+
+  const char8* garbage_begin = input;
```
Comment (strager) [inline src/lex.cpp:800]: Can we reuse the check_garbage_in_number_literal function here? If not, why not?

File: test/test-lex.cpp [hunk 24]
```diff
@@ -136,12 +136,119 @@ TEST(test_lex, lex_binary_numbers) {
...
+TEST(test_lex, DISABLED_fail_lex_binary_number) {
+  {
+    error_collector v;
+    padded_string input(u8"0b1.1");
+    lexer l(&input, &v);
+    EXPECT_EQ(l.peek().type, token_type::number);
+    l.skip();
+    EXPECT_EQ(l.peek().type, token_type::end_of_file);
```
Comment (strager) [inline test/test-lex.cpp:147]: Why should `0b1.1` parse as a single number token, but `0o52.2` parse as two number tokens? I think they should be consistent with each other. With your patch, they are consistent; `0b1.1` parses as two number tokens, and `0o52.2` also parses as two number tokens.

File: test/test-lex.cpp [hunk 25]
```diff
@@ -136,12 +136,119 @@ TEST(test_lex, lex_binary_numbers) {
...
+
+TEST(test_lex, lex_octal_numbers_strict) {
+  check_single_token(u8"000", token_type::number);
+  check_single_token(u8"001", token_type::number);
+  check_single_token(u8"00010101010101010", token_type::number);
+  check_single_token(u8"00010101010101010", token_type::number);
+  check_single_token(u8"051", token_type::number);
+  check_single_token(u8"0o51", token_type::number);
+  check_single_token(u8"0o123n", token_type::number);
+}
```
Comment (strager) [inline test/test-lex.cpp:165]: Possible edge cases we might want to test:

* `0o0`
* `0o0n`
* `0o01`

File: test/test-lex.cpp [hunk 26]
```diff
@@ -136,12 +136,119 @@ TEST(test_lex, lex_binary_numbers) {
...
+
+TEST(test_lex, lex_octal_numbers_lax) {
+  check_single_token(u8"058", token_type::number);
+  check_single_token(u8"058.9", token_type::number);
+}
+
+// TODO(👮🏾‍♀️)doesn't work
```
Comment (strager) [inline test/test-lex.cpp:172]: Remove this comment, since the test is working now?

File: test/test-lex.cpp [hunk 27]
```diff
@@ -136,12 +136,119 @@ TEST(test_lex, lex_binary_numbers) {
...
+TEST(test_lex, fail_lex_octal_numbers_lax) {
+  // Each scope is essentially `check-single-token`, but pulled out of the
+  // function to access the error_collector at the end
+  //
+  // idea: have `check_single_token` accept a callback with the errors
```
Comment (strager) [inline test/test-lex.cpp:177]: The helper function is implemented now. Use it! `check_tokens_with_errors`.

File: test/test-lex.cpp [hunk 28]
```diff
@@ -136,12 +136,119 @@ TEST(test_lex, lex_binary_numbers) {
...
+  check_single_token(u8"0o123n", token_type::number);
+}
+
+TEST(test_lex, lex_octal_numbers_lax) {
+  check_single_token(u8"058", token_type::number);
+  check_single_token(u8"058.9", token_type::number);
+}
+
+// TODO(👮🏾‍♀️)doesn't work
+TEST(test_lex, fail_lex_octal_numbers_lax) {
```
Comment (strager) [inline test/test-lex.cpp:173]: What does "lax" mean here?

File: test/test-lex.cpp [hunk 29]
```diff
@@ -136,12 +136,119 @@ TEST(test_lex, lex_binary_numbers) {
...
+      u8"0123n", {token_type::number},
+      [](padded_string_view input, const auto& errors) {
+        EXPECT_THAT(errors, ElementsAre(ERROR_TYPE_FIELD(
+                                error_unexpected_characters_in_octal_number,
+                                characters, offsets_matcher(input, 4, 5))));
+      });
+}
+
+// TODO (👮🏾‍♀️) (when strict mode implemented) tests to fail in
+// strict mode
```
Comment (strager) [inline test/test-lex.cpp:238]: This comment lacks context. Mention 'octal number literals', move the comment inside a test function, or delete the comment.

File: test/test-lex.cpp (hunk 10 above)
Comment (strager) [reply test/test-lex.cpp:246]: Let's remove this test from this commit. Instead, add a comment to issue #43.

File: test/test-lex.cpp [hunk 30]
```diff
@@ -196,13 +327,6 @@ TEST(test_lex, lex_invalid_big_int_number) {
                                 error_big_int_literal_contains_decimal_point,
                                 where, offsets_matcher(input, 0, 6))));
       });
-  check_tokens_with_errors(
-      u8"0123n", {token_type::number},
-      [](padded_string_view input, const auto& errors) {
-        EXPECT_THAT(errors, ElementsAre(ERROR_TYPE_FIELD(
-                                error_big_int_literal_contains_leading_zero,
```
Comment (strager) [inline test/test-lex.cpp:203]: I think error_big_int_literal_contains_leading_zero is likely more helpful than error_unexpected_characters_in_octal_number. Can we easily make the lexer continue to report error_big_int_literal_contains_leading_zero?

File: test/test-lex.cpp [hunk 31]
```diff
@@ -196,13 +327,6 @@ TEST(test_lex, lex_invalid_big_int_number) {
                                 error_big_int_literal_contains_decimal_point,
                                 where, offsets_matcher(input, 0, 6))));
       });
-  check_tokens_with_errors(
```
Comment (strager) [inline test/test-lex.cpp:199]: This change implies that the `number_begin[0] == u8'0'` check in `lexer::parse_number` will always be false. Clean up that dead code?

File: test/test-lex.cpp [hunk 32]
```diff
@@ -221,27 +345,25 @@ TEST(test_lex, lex_invalid_big_int_number) {
                 VariantWith<error_big_int_literal_contains_decimal_point>(_)));
       });
 
-  // Complain about both the decimal point and the leading 0 digit.
+  /* // Complain about both the decimal point and the leading 0 digit. */
```
Comment (strager) [inline test/test-lex.cpp:348]: \*/ wat /\*

File: test/test-lex.cpp [hunk 33]
```diff
@@ -221,27 +345,25 @@ TEST(test_lex, lex_invalid_big_int_number) {
...
   check_tokens_with_errors(
-      u8"01.2n", {token_type::number},
+      u8"01.2n", {token_type::number,token_type::number},
```
Comment (strager) [inline test/test-lex.cpp:350]: Hmm, I don't like this change, but if it's difficult to preserve this behavior, I'll tolerate it.

File: test/test-lex.cpp [hunk 34]
```diff
@@ -221,27 +345,25 @@ TEST(test_lex, lex_invalid_big_int_number) {
...
       [](padded_string_view, const auto& errors) {
         EXPECT_THAT(
             errors,
             UnorderedElementsAre(
-                VariantWith<error_big_int_literal_contains_decimal_point>(_),
-                VariantWith<error_big_int_literal_contains_leading_zero>(_)));
+                VariantWith<error_unexpected_characters_in_octal_number>(_)));
```
Comment (strager) [inline test/test-lex.cpp:355]: :+1:

File: test/test-lex.cpp [hunk 35]
```diff
@@ -221,27 +345,25 @@ TEST(test_lex, lex_invalid_big_int_number) {
...
       });
 
   // Complain about everything. What a disaster.
   check_tokens_with_errors(
-      u8"01.2e+3n", {token_type::number},
+      u8"01.2e+3n", {token_type::number,token_type::number},
```
Comment (strager) [inline test/test-lex.cpp:360]: Ditto line 350.

File: test/test-lex.cpp [hunk 36]
```diff
@@ -221,27 +345,25 @@ TEST(test_lex, lex_invalid_big_int_number) {
...
 
-  // Complain about both the decimal point and the leading 0 digit.
+  /* // Complain about both the decimal point and the leading 0 digit. */
   check_tokens_with_errors(
-      u8"01.2n", {token_type::number},
+      u8"01.2n", {token_type::number,token_type::number},
       [](padded_string_view, const auto& errors) {
         EXPECT_THAT(
             errors,
             UnorderedElementsAre(
```
Comment (strager) [inline test/test-lex.cpp:354]: Minor: `ElementsAre` produces better error messages than `UnorderedElementsAre`. Prefer the former where possible.

File: src/lex.cpp [hunk 37]
```diff
@@ -743,27 +776,52 @@ void lexer::parse_binary_number() {
...
+  if (is_0o) assert(strict);
+  char8* input = this->input_;
+
+  while (this->is_octal_digit(*input)) {
+    ++input;
+  }
+
+  if (!is_0o) {
+    switch (*input) {
+    QLJS_CASE_DECIMAL_DIGIT:
```
Comment (strager) [inline src/lex.cpp:789]: Could we use `is_decimal_digit` here?

File: src/lex.cpp [hunk 38]
```diff
@@ -743,27 +776,52 @@ void lexer::parse_binary_number() {
...
+      this->input_ = input;
+      this->parse_number();
+      return;
+    }
+  }
+
+  if (is_0o) switch (*input)
+    case u8'n':
+      ++input;
```
Comment (strager) [inline src/lex.cpp:798]: CmonBruh

**Review by strager — CHANGES_REQUESTED** [review]

**Comment by strager** [comment]
@vegerot Are you still interested in working on this task?

**Comment by vegerot** [comment]
> @vegerot Are you still interested in working on this task?

Yes I am.  Unfortunately I thought it would take less time than this.  Had I know how much this would be in addition to my other commitments I may have reconsidered.  I'm sorry

File: .gitignore (hunk 1 above)
Comment (vegerot) [reply .gitignore]: Sustained 👨🏻‍⚖️

File: src/lex.cpp (hunk 38 above)
Comment (vegerot) [reply src/lex.cpp:812]: Would you rather

```suggestion
  if (is_0o) switch (*input) case u8'n': ++input;
```

File: src/include/quick-lint-js/lex.h (hunk 18 above)
Comment (vegerot) [reply src/include/quick-lint-js/lex.h]: This is probably better.  Depends I think on how many cases there really are for octals in strict and sloppy* mode.  For example, is there a difference in `0o` in strict and sloppy mode?  Not as for as I know, but if there were it could result in a lot of enumeration.  Which...actually, is the point of an enum anyways.

* While trying to learn about how a [particular piece](https://github.com/v8/v8/blob/master/src/parsing/parser.cc#L1545) of V8 works, I found out they call non_strict mode "sloppy", which I like

File: src/lex.cpp (hunk 37 above)
Comment (vegerot) [reply src/lex.cpp:803]: Would this now be `parse_decimal_digits_and_underscores`?

File: test/test-lex.cpp (hunk 10 above)
Comment (vegerot) [reply test/test-lex.cpp:271]: [done](https://github.com/quick-lint/quick-lint-js/issues/43#issuecomment-707295300)

File: test/test-lex.cpp (hunk 24 above)
Comment (vegerot) [reply test/test-lex.cpp:171]: Agreed

File: test/test-lex.cpp (hunk 28 above)
Comment (vegerot) [reply test/test-lex.cpp:196]: um...probably nothing?  When I have `TEST(test_lex, lex_octal_numbers_lax)`, I was thinking of things that would pass in sloppy mode but not strict mode.  I don't think that logic makes sense here.  Since they'll fail either way

File: test/test-lex.cpp (hunk 30 above)
Comment (vegerot) [reply test/test-lex.cpp:229]: We'll see

File: src/lex.cpp [hunk 39]
```diff
@@ -756,27 +790,52 @@ void lexer::parse_binary_number() {
...
+    // both of which are invalid
     QLJS_CASE_DECIMAL_DIGIT:
-    QLJS_CASE_IDENTIFIER_START:
-      input += 1;
+      ++input;
       break;
+    QLJS_CASE_IDENTIFIER_START:
+    case u8'.':
+      ++input;
+      goto done_parsing_garbage;
```
Comment (strager) [inline src/lex.cpp]: Why do we stop at the first garbage character? This causes `0o123abcdef` to parse as `0o123a` (number) and `bcdef` (identifier) (with an error on `a`), right?

File: src/lex.cpp [hunk 40]
```diff
@@ -1273,7 +1332,8 @@ void lexer::skip_line_comment_body() {
       this->skip_whitespace();
       break;
     }
-    // TODO(strager): Should we handle null bytes differently for #! comments?
+    // TODO(strager): Should we handle null bytes differently for #!
+    // comments?
```
Comment (strager) [inline src/lex.cpp:1366]: Nit: Undo

File: test/test-lex.cpp [hunk 41]
```diff
@@ -161,13 +161,120 @@ TEST(test_lex, lex_binary_numbers) {
   check_single_token(u8"0B010101010101010", token_type::number);
 }
 
+// fails
```
Comment (strager) [inline test/test-lex.cpp:163]: Nit: Remove this comment. Preferably, leave this test for a separate commit. But I won't get mad if you leave the test here.

File: src/lex.cpp (hunk 22 above)
Comment (vegerot) [reply src/lex.cpp:750]: Moved to seperate commit

Comment (vegerot) [reply src/lex.cpp:750]: hey

File: src/lex.cpp [hunk 42]
```diff
@@ -737,14 +749,36 @@ const char8* lexer::end_of_previous_token() const noexcept {
 }
 
 void lexer::parse_hexadecimal_number() {
-  QLJS_ASSERT(this->is_hex_digit(this->input_[0]) || this->input_[0] == '.');
```
[the remark this answers is not in the document]
Comment (vegerot) [reply src/lex.cpp:740]: Moved to separate commit

File: test/test-lex.cpp (hunk 41 above)
Comment (vegerot) [reply test/test-lex.cpp]: Added more specific TODO

File: src/lex.cpp (hunk 40 above)
Comment (vegerot) [reply src/lex.cpp]: Done

File: src/lex.cpp (hunk 39 above)
Comment (vegerot) [reply src/lex.cpp]: Fixed

File: src/lex.cpp [hunk 43]
```diff
@@ -756,27 +790,52 @@ void lexer::parse_binary_number() {
...
+    QLJS_CASE_DECIMAL_DIGIT:
+      this->input_ = input;
+      this->parse_number();
+      return;
+    }
+  }
+
+  if (is_0o) switch (*input)
+    case u8'n':
+      ++input;
```
[the remark this answers is not in the document]
Comment (vegerot) [reply src/lex.cpp]: aoeu?

File: src/lex.cpp [hunk 44]
```diff
@@ -148,16 +151,25 @@ void lexer::parse_current_token() {
...
       case 'b':
       case 'B':
         this->input_ += 2;
         this->parse_binary_number();
         break;
+      case 'o':
+        this->input_ += 2;
+        this->parse_octal_number(true, true);
+        break;
+        // probably wrong
```
[the remark this answers is not in the document]
Comment (vegerot) [reply src/lex.cpp]: Sorry.  That comment was probably wrong.

File: src/lex.cpp (hunk 38 above)
Comment (vegerot) [reply src/lex.cpp]: 4Head
